# **Python 0: Introduction**

> - [A] Jupyter notebook cells
> - [B] The print function
> - [C] Arithmetic operators
> - [D] Variables
> - [E] Comparison operators
> - [F] Conditional statements
> - [G] Packages and modules


---
## **[A] Jupyter notebook**

This Jupyter notebook contains two types of **cells**

> - **text cell**: The text you are reading right now is placed in a **text cell**. <br> These cells do not contain Python code, but instead contain text and instructions
> - **code cell**: Below this text cell we find the first **code cell**. <br> A code cell has a gray background and it shows `[ ]` in the left margin

Code cell below:


In [ ]:
2 + 2

4

The code cell needs to be **filled** with Python code to be useful.

> **DO-IT**
1. Click on the code cell to activate it, indicated by a blinking cursor
2. Once the cell is activated, you can start writing Python code
3. Try typing in `2 + 2`

The code cell needs to **run** to execute its contents. There are several ways to do this in a Jupyter notebook.

> **DO-IT**
- If the code cell is not activated (not highlighted):
  - Hover over `[ ]` and click on the "play" icon
- If the code cell is activated (is highlighted)
  - Run the code cell: Click on the "play" icon
  - Run the code cell: Ctrl+Enter (Command+Enter on a Mac) <br>(Note: This is my *preferred* way, as it does not shift focus to the next cell)
  - Run the code cell and move to the cell below: Shift+Enter
  - Run the code cell and create a new code cell below: Alt+Enter (Option+Enter on a Mac)
  - **I strongly advise you _not_ to use the "play" icon, using your keyboard is much faster**


After the code cell has run its **output** is displayed directly below the cell.

> There are a few things to consider here:
-  After a code cell has run for the first time, the square brackets `[ ]` are replaced by `[1]`
- The number between the square brackets is the **execution number** (order) of the code cell
- If the code cell is executed again, this number increments with 1: [1] becomes [2], and so on and so forth

In the rest of this notebook I will indicate the code cells in which you have to program using a **comment**

> **Comment**: text imbedded in your Python code that has no functionality

- Comments are used by programmers to clarify their code
- To create a single-line comment in Python use a `#` (called the "hash" symbol or the "pound" symbol)
- If I want you to program in a particular code cell, it will contain the comment: `#DO-IT`, like below

In [ ]:
# DO-IT

---
## **[B] The print function**

The canonical way to get introduced to any programming language is by writting a program that outputs "Hello, World!" to the screen.

> - In programming, **outputting** something to the screen is known as **printing**
> - In Python, you **print** something to the screen by *calling* (using) the **print function**

In [ ]:
print('Hello, World!')

SyntaxError: invalid syntax (<ipython-input-10-0eb003ed9968>, line 1)

Let's dissect this line of code, piece by piece:
- **`print`** is the **name** of the function we are calling
- **`()`** are **parentheses** (sometimes called round brackets) <br>
  - The parentheses following a function name can be filled with **arguments**
- **`'Hello, World!'`** is the single **argument** we provide to the print function <br>
  - Think of arguments as *inputs* to a function

Note that the argument **`'Hello, World!'`** is surrounded by single quotes **`' '`**
> - In this case, the argument **`'Hello, World!'`** is called a **string** <br>
A **string** is a sequence of characters, i.e. a text snippet
> - Python needs the single quotes **`' '`** to distinguish text from code
> - In Python, you can use either single quotes **`' '`** or double quotes **`" "`** to indicate a string. There is no real difference, it is a matter of preference
> - Note that **the quotes are not printed** as they are not part of the string, they are only used to indicate to Python that we are providing a string

Note: Python is very versatile in handling strings (that is, text data), it is one of its strengths!

> **DO-IT**
- In the code cell below, print the *Hello, World!* string using double quotes instead of single quotes
- Is the output of the print function the same as when you used single quotes?

In [ ]:
#DO-IT
print("Hello, World!")
print('Hello, World!')
# yes, output is the same

Hello, World!
Hello, World!


The print function can accept more than one argument
> - Multiple arguments for a function are separated by a **`,`** (comma)

In [ ]:
print('H', 'e', 'l', 'l', 'o', ',', 'World!')

H e l l o , World!


Note that the print function automatically **separates** the **output** of arguments with a space. This is a property of the print function and has nothing to do with the spaces between the arguments in your code.

> **DO-IT**
> - Try adding and removing spaces *between* the arguments in your print function call
> - Does the output of the print function change?

In [ ]:
#DO-IT
print('H', 'e',
      'l',
      'l',
      'o', ',', 'World!')
print('H','e','l','l','o',',','World!')
print('H',    'e',    'l',    'l',    'o',    ',',    'World!')
# output does not change

H e l l o , World!
H e l l o , World!
H e l l o , World!


Note: Within a  single code cell you can have multiple function calls:

In [ ]:
print('Hello')
print(',')
print('World!')

Hello
,
World!


---
## **[C] Arithmetic operators**

In data science and statistical computing **arithmetics** with numbers are essential.

In Python there are several arithmetic operators that operate on numerical values:
> * `+`
> * `-`
> * `*`
> * `/`
> * `**`
> * `%`
> * `//`

You are probably familiar with the first four arithmetic operators, but the last three operators may be new to you.

> **DO-IT**
> * Add 2 and 3 together

In [ ]:
#DO-IT
2 + 3

5

> **DO-IT**
> * Subtract 12 from 9

In [ ]:
#DO-IT
9 - 12

-3

> **DO-IT**
> * Multiply 13 times 7

In [ ]:
#DO-IT
13 * 7

91

> **DO-IT**
> * Divide 42 by 27

In [ ]:
#DO-IT
42 / 27

1.5555555555555556

In the `print` function we have seen that we use parentheses to collect the arguments for a function.

> In Python parentheses are **also** used to indicate the **order of operations** in arithmetic operations.
<br>
> !!! This is important !!! **Parentheses have multiple functions in the Python syntax**.

Most likely you have used parentheses like this on a calculator before.

> You can find the standard calculation rules about the order of operations here: https://en.wikipedia.org/wiki/Order_of_operations

> **DO-IT**
> * In *one line of Python code*:
>  * First add 3 and 4 together
>  * Then multiply the result times 5
> * Verify that your answer is correct!

In [ ]:
#DO-IT
(3 + 4) * 5

35

---
## **[D] Variables**


More often than not, we want to store our results in the computer's memory. In Python, we use a **variable** to store a result.

For example, we can store the result of 2 + 2 as follows:



In [ ]:
two_plus_two = 2 + 2

Let's dissect this line of code, piece by piece:
- **two_plus_two** is the **name** of the variable in which we are storing the result
- The **`=`** operator tells Python that we want to assign the **value to the right** of the **`=`** to the variable **name to the left** of the **`=`**

Note: Python did not display the result of the computation.

> Does this mean that Python did not do the computation? **No**
> 1.   Python first performs the computation: `2 + 2`
> 2.   The result (`4`) is assigned to the variable name `two_plus_two`

If we now want to see the result, we can ask Python to display or print the variable `two_plus_two`

In [ ]:
print(two_plus_two)

4


It is a good idea to use an expressive **variable name** to make the intent of your code more clear.

When choosing a variable name, use a combination of the following **valid characters**:
> *    Uppercase letters: A, B, ..., Z
*    Lowercase letters: a, b, ..., z
*    Numbers:  0, 1, 2, ..., 9 (**Note: A variable name cannot start with a number!**)
*    Underscore: _

In addition, names in Python (variable, function) are **case sensitive**

> **DO-IT**
*    Is `TWO_PLUS_TWO` a valid variable name? Why, or why not?
*    Is `2+2` a valid variable name? Why, or why not?
*    Is `2_plus_2` a valid variable name? Why, or why not?
*    Is `2 plus 2` a valid variable name? Why, or why not?

In [ ]:
#DO-IT
TWO_PLUS_TWO = 2 + 2 # is fine
#2+2 = 2 + 2 # not fine, variable name cannot contain a + sign
# 2_plus_2 = 2 + 2 # not fine, variable name cannot start with a number
# 2 plus 2 = 2 + 2 # not fine, cannot contain spaces

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (<ipython-input-23-bcdb7b4f1dad>, line 3)

In Python a variable can hold any type of value. We have just seen the results of a numerical computation, but it can also hold a **string**

> **DO-IT**
*    In the code below, assign your own name to the variable called **`my_name`**

In [ ]:
#DO-IT
my_name = 'XH'
print('My name is:', my_name)

My name is: XH


### **Viewing variables in Jupyer Notebook**

We have now created a few variables in our Python session. In particular in longer Jupyter notebooks, it may be difficult to keep track of all these variables.

Fortunately, Google Colab provides a **Variables tab** that gives us an overview of our variables. It is located at the **left-hand side** of the screen and looks like: $$\{x\}$$

For each variable it displays the:
- Name
- Type
- Shape
- Value

Type/Shape/Value will be discussed later.

If we create a new variable, the Variables tab will be automatically updated:

In [ ]:
new_variable = "Something new"
type(new_variable)

str

## **[E] Comparison operators**

Besides arithmetics, one of the corner stones of programming is comparing values to each other. For example, you want to compare whether two numbers are equal, or whether one number is larger than another.

In Python there are several comparison operators:
> * Equality: `==`
> * Inequality: `!=`
> * Less than: `<`
> * Less than or equal: `<=`
> * Greater than: `>`
> * Greater than or equal: `>=`

The result of these comparison operators is either **`True`** or **`False`**.

> A value that is **`True`** or **`False`** is called a **`boolean`**

This is another variable type, just like the numbers and strings we have seen before.

### **Equality: `==`**

Use the **`a == b`** syntax:

In [ ]:
a = 3.14
b = 744

print(a == b)
print(a == a)

False
True


### **Inequality: `!=`**

Use the **`a != b`** syntax:

In [ ]:
a = 3.14
b = 744

print(a != b)
print(a != a)

True
False


### **Less than: `<`**

Use the **`a < b`** syntax:

In [ ]:
a = 3.14
b = 744

print(a < b)
print(a < a)

True
False


### **Less than or equal: `<=`**

Use the **`a <= b`** syntax:

In [ ]:
a = 3.14
b = 744

print(a <= b)
print(a <= a)

True
True


### **Greater than: `>`**

Use the **`a > b`** syntax:

In [ ]:
a = 3.14
b = 744

print(a > b)
print(a > a)

False
False


### **Greater than or equal: `>=`**

Use the **`a >= b`** syntax:

In [ ]:
a = 3.14
b = 744

print(a >= b)
print(a >= a)

False
True


---
## **[F] Conditional statements**

A basic computer program executes each line, after line, after line, ..., in a predefined order.

> **Control flow** allows you to change the order in which your computer program is executed
<br>
> The order is now made conditional on values of your variables

Today we will first look at **conditional statements**, but later on during the course we will explore **`for`** loops and **`while`** loops as well.

### **`if`**

A block of code under **`if STATEMENT:`** syntax is only executed if the **`STATEMENT`** is **`True`**

> * Note the **indentation** (whitespace) of the block under the **`if`** statements!
> * You create an indentation by hitting the **tab** button **inside** the code cell
> * This use of whitespace is one of the more unique features of Python
> * Compare this against R, where you use curly braces `{ }` to indicate a block of code

In [ ]:
x = 3
y = 4

if x == y:
  print('x and y the same')
  print('x=y:', x)

if x != y:
  print('x and y are not the same')
  print('x:', x)
  print('y:', y)

As a special case we can use **numerical variables** directly in a conditional statement.

> Let **`x`** be a number
> * If **`x`** is **nonzero** it is considered **True** in a conditional statement
> * If **`x`** is **zero** it is considered **False** in a conditional statement
> * In essence, **`if x:`** is equavalent to **`if x != 0:`**

In [ ]:
x = 5
if x:
  print('This prints if x is not equal to zero:', x)

y = 0
if y:
  print('This does not print if y is equal to zero:', y)

z = 3.14
if z:
  print(z)

This prints if x is not equal to zero: 5
3.14


### **`if`**, **`else`**

The **`if STATEMENT:`** syntax can be extended using an **`else:`** statement.

The block of code under the **`else:`** will only be executed if **`STATEMENT`** is **`False`**:

In [ ]:
course_number = 123

if course_number == 744:
  print('My favorite course is:', course_number, '(but you knew that already)')
else:
  print(course_number, 'is definitely a fun course, but not my favorite')

123 is definitely a fun course, but not my favorite


### **`if`**, **`elif`**, **`else`**

The **`if STATEMENT:`** syntax can be further extended by using an **`elif OTHER_STATEMENT:`** statement.

The block of code under the **`elif OTHER_STATEMENT:`** will only be executed if **`STATEMENT`** is **`False`** and **`OTHER_STATEMENT:`** is **`True`**:

The block of code under the **`else:`** will only be executed if both **`STATEMENT`** and **`OTHER_STATEMENT:`** are **`False`**:

> About **`elif`**:
> - It is always placed *after* **`if`**, but *before* an **`else`**
> - You can have *more than one* **`elif`**

In [ ]:
course_number = 123

if course_number == 744:
  print('I teach', course_number, 'in the MSMA program')
elif course_number == 452:
  print('I teach', course_number, 'in the UG program')
elif course_number == 858:
  print('I teach', course_number, 'in the PhD program')
else:
  print('I do not teach', course_number)

I do not teach 123


---
## **[G] Packages and modules**

**Packages** (also known as modules) *significantly* extend the functionality of Python

> *A Python package (module) is similar to an R package*

Two types of packages in Python:
* **Internal**: Always available (also called built-in modules)
* **External**: Separate packages that (sometimes) need to be installed

Examples of popular (external) data science packages:
> * `numpy`
> * `pandas`
> * `scipy`
> * `scikit-learn`

In this course you will become familiar with the packages mentioned above.
> * Although these packages are external, they come pre-installed on Google Colab. <br> **You do not have to install anything yourself!**

### Loading a Package

A package can be loaded using the **`import`** statement. After a package has been *imported* you can access its functionality.

There are **4 options to importing a package**:
> 1. Explicit import of a package
> 2. Explicit import of a package by alias
> 3. Explicit import of **some** of a package's contents
> 4. *Implicit* import of **all** of a package's contents

Only options 1 and 2 should be used. Option 3 is fine, but is somewhat of limited use in practice. **Option 4 should not be used**.

### 1. Explicit import of package
Using the **`import PACKAGE`** syntax:


In [ ]:
import math

After the `math` module has been explicitly imported, you can access its contents using `math.CONTENT`.

Example 1:
> The square root function `sqrt()` is now available through the `math` module

In [ ]:
math.sqrt(9)

3.0

Example 2:
> The value of pi (3.1415...) `pi` can be accessed through the `math` module

In [ ]:
math.pi

3.141592653589793

### 2. Explicit import of a package by alias
Using the **`import PACKAGE as ALIAS`** syntax:

In [ ]:
import math as m

The contents of the `math` module can now be accessed using `m.CONTENT`.

> The `sqrt()` function in `math` can now be accessed as `m.sqrt()`

In [ ]:
m.sqrt(16)

4.0

### 3. Explicit import of **some** of a package's contents
Using the **`from PACKAGE import CONTENT_A, CONTENT_B`** syntax:

In [ ]:
from math import log, pi, e

In the cell above we have imported the following contents from the `math` module:
* Function `log()` (natural logarithm)
* Values `pi` and `e` (Euler's number)

This content can now be accessed directly, without having to prepend with `math.` or `m.`


In [ ]:
print(pi)
print(log(744))
print(log(e))


3.141592653589793
6.612041034833092
1.0


NameError: name 'math' is not defined

### 4. Implicit import of **all** of a package's contents (**not recommended!**)
Using the **`from PACKAGE import *`** syntax:

In [ ]:
from math import *

Here the asterisk (`*`) means that all the content of the `math` module can now be accessed directly, without prepending with `math.`

In [ ]:
print(cos(1.0))
print(sin(1.0))
print(sqrt(744))

0.5403023058681398
0.8414709848078965
27.27636339397171


**Note: This is really bad programming style and strongly discouraged!!!**

# **Python 1: More Python syntax**

> - [A] Objects, types, and the ``type`` function:

Make sure you understand how behavior of a variable changes if you are dealing
 with a different **variable type**.
> - [B] Built-in data structures (tuple and set excluded).
> - [C] Slicing: Completely excluded on the quiz (but you should know how to slice a numpy array)
> - [E] Creating and using functions:


Make sure that you understand how to work with **positional** arguments and **keyword** arguments when calling a function.

---
## **[A] Objects, types, and the ``type`` function**

In Python, everything is a so-called **object**. There are different programming paradigms, and **object-oriented** programming is just one of them.

Conceptually, an object is an entity that has a:
> - *class*: think of this as the **type** of the object
>  - This class is important! An object's class determines **what you can do** with your object
> - *state*: think of this as the **data** related to the object<br>
>  - In Python these are called the object's **attributes**
> - *behavior*: think of this as the **functionality** associated with the object's **type**<br>
>  - In Python these are called the object's **methods**
>  - One way to think of a **method** is that it is a function specifically designed for the **object**

If you have an object stored in **`variable_name`**, you can access it's method named **`method_name`** as follows:
>  **`variable_name.method_name()`**
>  - Note 1: Any arguments to the method should be placed within the parentheses
>  - Note 2: Sometimes a **method** takes no arguments (empty parentheses), while other times it does

Let us explore this with two types that we have seen before, an **integer** and a **string**:

In [ ]:
course_code = 744
course_name = 'BUMK'

We can **determine the class** (think: type) of these objects (variables) using the **`type`** function:

In [ ]:
print(type(course_code))
print(type(course_name))

<class 'int'>
<class 'str'>


This tells us that:
> - `course_code` is of type **`int`**, which stands for **integer**
> - `course_name` is of type **`str`**, which stands for **string**

Again:
> **It is important to know what the class (type) of your variable is:<br>It determines what you can do with your variable**

Let me illustrate that in the next cell:

> **DO-IT**<br>
> Note: Getting errors for some of these is OK!<br>Read the errors and try to understand them, they are informative!
1. Divide `course_code` by 2
2. Divide `course_name` by 2
3. Multiply `course_code` by 2
4. Multiply `course_name` by 2
5. Add the **number** 42 to `course_code`
6. Add the **number** 42 to `course_name`
7. Add the **string** '42' to `course_code` (Notice the single quotes marks!)
8. Add the **string** '42' to `course_name` (Notice the single quotes marks!)


In [ ]:
#DO-IT
#course_name / 2
#course_name * 2
#course_name + 42
#course_name + '42'
#course_code + '42'

'BUMK42'

A few things to take away here:

> - Some operations are allowed on an **`int`** but not on a **`str`**, and vice versa
> - Python is more versatile in handling strings than you expected (probably)!

The **`str`** type actually has *a lot* of **methods** that can be accessed using the **`variable_name.method_name()`** syntax.
> - We will look at the `lower`, `capitalize`, `isupper`, and `islower` methods related to the **`str`** type
> - More **`str`** methods can be found on the relevant page in the documentation:
> - https://docs.python.org/3/library/stdtypes.html#string-methods

> **DO-IT**<br>
> On the **`course_name`** variable, use the following methods and print the results:
1. `lower`
2. `capitalize`
3. `isupper`
4. `islower`
> * Note: Make sure you understand what these methods are doing
> * Note: The Python manual is a useful resource!

In [ ]:
#DO-IT
#course_name
# course_name.lower()

print(course_name)
print(course_name.lower())
print(course_name.capitalize())
print(course_name.isupper())
print(course_name.islower())

BUMK
bumk
Bumk
True
False


### The **`None`** type
There is a special type in Python that needs some explanation: **`None`**
> Think of `None` as the **empty/unspecified type** (this is a loose definition)

`None` most commonly shows up in Python code after calling a function that does not explicitly return a value
- **We have already seen such a function**: `print`
- `print` only displays (prints) something to the screen, **it does not return any value**
- Therefore, you cannot assign the output of `print` to a variable

> **DO-IT**<br>
1. Determine the type of `print('BUMK744')`
2. Store the value of `print('BUMK744')` in a variable called `print_result`, print that variable, and determine its type

In [ ]:
#DO-IT

# 1.
print(type(print('BUMK744')))

BUMK744
<class 'NoneType'>


In [ ]:
# 2.
print_result = print('BUMK744')
print(print_result)
print(type(print_result))

BUMK744
None
<class 'NoneType'>


---
## **[B] Built-in data structures**

Think of a **data structure** as a **container** that can hold multiple variables, possibly of mixed types.
* A simple conceptual example is a spreadsheet (e.g. with rows and columns)
* A spreadsheet is a **data structure** because it *contains* numbers (or text, or both).

We can compare some Python data structures to those in the R programming language:
- The Python equivalent of the **data.frame** in R is the **DataFrame** type from the external **pandas** module
- The Python equivalent of the **matrix** in R is the **numpy array** type from the external **numpy** module

Note: We will discuss these more advanced data structures later in this course.

For now, we start with the **built-in** data structures
* These are much simpler to start with
* They do not rely on any external modules (they are part of the standard library)

Python has four types of built-in data structures:
> - **`list`**:
>  - *ordered* collection of elements
> - **`tuple`**:
>  - *ordered* collection of elements that is **immutable** (immutable: cannot be changed after creation)
> - **`dict`** (dictionary):
>  - *unordered* dictionary of (key, value) pairs
>  - Note: Since Python version 3.7 a `dict` is actually **ordered**. <br> However, it is best **not to rely** on this new property for backwards compatibility with older Python versions
> - **`set`**:
>  - *unordered* collection of **unique** elements

### **`list`**

A list is an **ordered** data structure. This means:
> - The **order in which the values are stored matters**
> - You access values in a list by using an **index** number

The syntax of **constructing a list**
> - Use **square brackets**
> - **Separates values** in the list using a **comma**

In [ ]:
f = [2, 3, 5, 7, 11]
print(f)
print(type(f))

[2, 3, 5, 7, 11]
<class 'list'>


The number of elements is also called the **length of the list**
> - Length can be accessed using the **`len`** function
> - Note: **`len`** is a general function and specifically  **not a method of the list class**. <br> Therefore it does **not** use the `object_name.method_name()` syntax.<br> (I know, this is confusing!!)

In [ ]:
print(len(f))
# print(len(course_name))

len(f)

5


AttributeError: 'list' object has no attribute 'len'

You can access a **single value** in a list by **indexing** the list using an **index number**

> - The index **`0`** refers to the **first element** in the list
> - The index **`1`** refers to the **second element** in the list
> - ...

**!! IMPORTANT !!** Indexing in Python **starts at zero (0)**, not at one (1)

> **DO-IT**<br>
1. Select the first element from the list `f`
2. Select the fourth element from the list `f`

In [ ]:
#DO-IT
print(f)
# 1.
print(f[0]) # 0 refers to the 1st element
# 2.
print(f[3]) # 3 refers to the 4th element

[2, 3, 5, 7, 11]
2
7


**Values at the back of the list** can be accessed using **negative index numbers**:
> - The index **`-1`** refers to the **last element** in the list
> - The index **`-2`** refers to the **second-to-last element** in the list
> - ...

> **DO-IT**<br>
1. Select the **last** element from the list `f` using a **positive index number**
1. Select the **last** element from the list `f` using a **negative index number**
2. Select the **first** element from the list `f` using a **negative index number**

When is it easier to use positive and negative index numbers?

In [ ]:
#DO-IT
print(f)
# 1.
print(f[4]) # this is the last element in the list
# 2.
print(f[-1]) # this is the last element in the list
# 3.
print(f[-5]) # this is the first element in the list

[2, 3, 5, 7, 11]
11
11
2


A list is **mutable**: This means that its **values can be updated** after creation:

In [ ]:
print(f)
f[0] = 744
print(f)

[2, 3, 5, 7, 11]
[744, 3, 5, 7, 11]


Furthermore, a list is a very flexible container and can contain **mixed types**:

In [ ]:
g = ['this is a string', 744, 3.14, True, False, f]
print(g)

['this is a string', 744, 3.14, True, False, [744, 3, 5, 7, 11]]


Note: The list `g` contains another list `f`!

Multiple lists can be **concatenated** (added together) using the **`+`** operator:

In [ ]:
[1, 2, 3] + [4, 5] + [6]

[1, 2, 3, 4, 5, 6]

Just like the **`str`** type, the **`list`** type also specifies different **methods**.

> - We will look at the `append` and `count` methods related to the **`list`** type
> - More **`list`** methods can be found in the documentation: <br> https://docs.python.org/3/tutorial/datastructures.html

In [ ]:
print(g)
g.append('This is new')
print(g)
print(g.count(True))
print(g.count('True'))

['this is a string', 744, 3.14, True, False, [744, 3, 5, 7, 11], 'This is new']
['this is a string', 744, 3.14, True, False, [744, 3, 5, 7, 11], 'This is new', 'This is new']
1
0


Note: The `append` method changes the list!
> If you run the code cell multiple times, the string `'This is new'` will be added to the list multiple times
<br>
> Remember: We discussed the importance of the execution order last lecture

### dict
A dictionary is a **mapping** of **keys** to **values**

**keys**
- Think of a **key** as an index, such as we have seen for a list and a tuple
- The difference is that the keys of a dictionary are specified by you: they are not determined by the order of entry.

**values**
- A **value** in a dictionary can be accessed by using its corresponding **key**

Some technical detail (not important)
> A dictionary key value should be **hashable**. Amongst others, strings and numbers are hashable, so in this course we will use these types as dictionary keys.

The syntax of constructing a dictionary uses **curly braces** and each **key: value** pair is separated by a comma:

In [ ]:
d = {'code': 744, 'name': 'BUMK', 42: 'The key here is a number'}
print(type(d))

<class 'dict'>


Values in a dictionary are accessed in a similar way as a list and a tuple. The difference is that **a dictionary's index are the keys** corresponding to the values, instead of an index number that starts at 0 such as for lists and tuples.

In [ ]:
print(d)
print(d['code'])
print(d[42])

# # this will error, because 0 is not a valid key in our dictionary
# print(d[0])

{'code': 744, 'name': 'BUMK', 42: 'The key here is a number'}
744
The key here is a number


You can add new values to a dict `d` using the following syntax:
```python
# adding a new_value to the dictionary that is located using new_key
d[new_key] = new_value
```

In [ ]:
d['N'] = 64

In [ ]:
print(d)

{'code': 744, 'name': 'BUMK', 42: 'The key here is a number', 'N': 64}


Just as the other types the **`dict`** type has its own collection of methods. However, these methods are best explained once we get to **looping**.

---
## **[E] Creating and using functions**

We have worked with functions before, such as the `print` function. As a reminder, this is a function call to `print`:
```python
print('BUMK744')
```
Here:
> - `print` is the **function name**
> - `( )` are **parentheses for calling a function**
> - `BUMK744` is the **argument** to the function


In Python you can define your own function using the following syntax:
```python
def function_name(func_arg_0, func_arg_1):
    # code of your function
    # Notice that this code should be indented!
```

> **DO-IT**<br>
1. Define a function called `first_function`
2. It should take two arguments named `greeting` and `name`
3. The function should print `greeting` and `name` using a single call to the `print` function, and then finally end with a double exclamation mark (!!)
4. Finally, call your function to make sure that it works as expected

In [ ]:
#DO-IT
def first_function(greeting, name):
  print(greeting, name + '!!')

first_function('Hi', 'Bob')

Hi Bob!!


If we call a function by just providing values for the function's arguments (`greeting` and `name` in the above example), without using the names of the function's arguments, these are called **positional arguments**.

Another way to provide arguments to a function in Python is by explicitly supplying the name of the argument to which the value corresponds. This is called a **keyword argument** and can make your code more clear:

In [ ]:
first_function(greeting='Hello', name='John')

Hello John!!


In fact, **you can change the order of arguments if you are using keyword arguments**, without changing the function's behavior:

In [ ]:
first_function(name='John', greeting='Hello')

Hello John!!


You can also mix positional and keyword arguments, but in that case the keyword arguments should always come at the end of the function call. For example:

```python
# OK
first_function('Hi there', name='John')

# NOT OK
first_function(greeting='Hi there', 'John')
```

In [ ]:
#TRY-IT


A good example of this is if your function has some default argument that you want to override. For example, consider the `print` function. By default, the arguments of the `print` function are separated using a single space. You can override this behavior by changing the **`sep`** keyword argument of the `print` function.

> **DO-IT**<br>
1. Change the behavior of the `print('Hi', 'Bob')` call by adding the **`sep`** keyword argument to change the separator from a single space to `*-*-*`

In [ ]:
#DO-IT
# 1.
print('Hi', 'Bob', sep='*-*-*')

Hi*-*-*Bob


# **Python 2: NumPy**

> - [A] NumPy arrays

Be familiar with how to **select data and slice data** like a row or a column from a numpy array as well. Only basic slicing is required.
Make sure you understand the **axis** argument that we can supply to a 2-dimensional numpy array and how that gives you results per column (axis=0) or per row (axis=1).

> - [C] Random sampling


Make sure you understand how to control **(pseudo) random numbers**.

---
## **[A] NumPy**



In most Python code `numpy` is imported under the *alias* `np`, we will follow that convention as well:

In [ ]:
import numpy as np
#import numpy
#from numpy import array
#from numpy import *

After importing NumPy, all its functionality is available to use in your own Python programs.

- We will start working with the **numpy array**
- This is the datatype associated with NumPy and can be compared to **R's matrix datatype**.

### NumPy array creation

To create a numpy array you use the **`np.array()`** function. You can call it directly as follows:

In [ ]:
v = np.array([5, 3.14, 7])
print(v)
print(type(v))

[5.   3.14 7.  ]
<class 'numpy.ndarray'>


Or alternatively, you can create a numpy array from an existing list (or a tuple):

In [ ]:
list_or_tuple = [1, 2, 3]
v_alt = np.array(list_or_tuple)
print(v_alt)
print(type(v_alt))

[1 2 3]
<class 'numpy.ndarray'>


The types used in a numpy array should be **homogenous** (generally speaking)
- This means that the **elements of a numpy array should have the same type**.
- **Note**: Theoretically you can mix different types in a numpy array but that is not the intended use of numpy. If you want to mix different types use a pandas DataFrame instead (next lecture).

A numpy array that only consists of a single row (or column), as we have defined above, is known as a **vector**.

We can also create a numpy array that consists of multiple rows and multiple columns. This is called a **matrix** (think of a table, or a spreadsheet).

For example, the matrix containing the values:
```
11, 12
21, 22
31, 32
```
can be created as follows:

In [ ]:
m = np.array([
    (11, 12), # a comma WITHIN parentheses is used to separate values on a row
    (21, 22), # a comma BETWEEN parentheses is used to separate rows
    (31, 32), # this last comma is actually optional (but only the last one!)
])

print(m)

[[11 12]
 [21 22]
 [31 32]]


A few remarks:
- In the creation of a numpy array, the use of whitespace and multiple lines is actually not necessary. It is just a notational convenience and can make your code easier to read.
- In the argument of `np.array` you could replace the parentheses by square brackets, and vice versa. However, calling the function `np.array` **always** require parentheses. In other words, the outer pair of parentheses always should be a pair of parentheses

> **DO-IT**<br>
1. *To practice with creating a matrix yourself*<br>Create a matrix called `m1` with the following values and print it:
```
1, 2, 3, 4
5, 6, 7, 8
```
2. *To show you that whitespace and linebreaks do not matter*<br>Create a matrix called `m2` with the following values and print it:
```
1, 2, 3, 4
5, 6, 7, 8
```
However, this time, **place everything on a single line of Python code**
3. *To show you that you can swap parentheses and square brackets*<br>Create a matrix called `m3` with the following values and print it:
```
1, 2, 3, 4
5, 6, 7, 8
```
However, this time, **only use parentheses** in the argument to `np.array`
3. *To show you that you can swap parentheses and square brackets*<br>Create a matrix called `m4` with the following values and print it:
```
1, 2, 3, 4
5, 6, 7, 8
```
However, this time, **only use square brackets** in the argument to `np.array`

In [ ]:
# DO-IT

# 1.
m1 = np.array([
  [1, 2, 3, 4],
  [5, 6, 7, 8],
])
print(m1)

# 2.
m2 = np.array([(1, 2, 3, 4), (5, 6, 7, 8),])
print(m2)

# 3.
m3 = np.array(((1, 2, 3, 4), (5, 6, 7, 8)))
print(m3)

# 4.
m4 = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])
print(m4)

[[1 2 3 4]
 [5 6 7 8]]
[[1 2 3 4]
 [5 6 7 8]]
[[1 2 3 4]
 [5 6 7 8]]
[[1 2 3 4]
 [5 6 7 8]]


### NumPy array attributes

We can use the **attributes** of the numpy array to get information about its dimensions.

A few important remarks here:
- An attribute is not the same as a method (read: function).
- An attribute is part of the **state** (read: data) of the object (read: numpy array)
- Therefore **using parentheses when using an attribute does not work**

The first attribute we will look at is the **`shape`** attribute

**`shape`**:
> - **`shape`** returns a tuple with the number of rows and columns of the numpy array
> - That is: It returns (number of rows, number of columns)

In [ ]:
print(v.shape)
print(m.shape)

(3,)
(3, 2)


**`ndim`**:
> - **`ndim`** returns the number of dimensions of the numpy array.
> - If the numpy array is a vector this will be equal to 1
> - If the numpy array is a matrix it will be equal to 2
> - In fact, numpy arrays can be of an even higher dimension as well

**DO-IT**<br>
1. Use `ndim` to print the number of dimensions for the **`v`** variable
2. Use `ndim` to print the number of dimensions for the **`m`** variable

In [ ]:
#DO-IT

# 1.
print(v.ndim)

# 2.
print(m.ndim)

1
2


**`size`**:
> - **`size`** returns the total number of elements in the numpy array
> - That is: number of rows * number of columns

**DO-IT**<br>
1. Use `size` to print the number of elements for the **`v`** variable
2. Use `size` to print the number of elements for the **`m`** variable

In [ ]:
#DO-IT

# 1.
print(v.size)

# 2.
print(m.size)

3
6



Note:
- We have used the `len` function before to determine the length of a list, for example
- We can also use the `len` function on a numpy array, but the results may surprise you at first.

**DO-IT**<br>
1. Use the `len` function and print its output for the **`v`** variable
2. Use the `len` function and print its output for the **`m`** variable
3. Can you figure out what the `len` function returns?<br>Hint: Create new variables (both vectors and matrices) with a different number of rows and columns


In [ ]:
#DO-IT

# 1.
print(len(v))

# 2.
print(len(m))

# 3.
# Answer:   The len function returns the number of elements in the FIRST
#           dimension of a numpy array.
#
#           For v, we only have one dimension - so that is also the first
#           dimension - and that dimension has 3 elements. Therefore, len(v)=3.
#
#           For m, we have two dimensions, but len only returns the number of
#           elements in the FIRST dimension. As we have 3 rows, len(m)=3.

3
3


**`dtype`**:
> - **`dtype`** returns the type of the elements in the numpy array
>  - `float`: If the array only contains floats (fractional numbers)
>  - `int`: If the array only contains integers (whole numbers)
> - If you see something like `float64` or `int64` that means numbers on your computer are stored in 64-bits. This is the default on most modern computers.

**Note**: Generally speaking it is more useful to consider the `dtype` of a NumPy array then its type

**DO-IT**<br>
1. Use `dtype` to print the type of the elements for the **`v`** variable
2. Use `dtype` to print the type of the elements for the **`m`** variable

In [ ]:
#DO-IT

# 1.
print(v.dtype)

# 2.
print(m.dtype)

float64
int64


### Accessing values

Accessing values in a **vector** numpy array works the same as with lists:

In [ ]:
print(v)
print(v[0]) # first ELEMENT
print(v[-1]) # last ELEMENT
print(v[1:]) # slicing

[5.   3.14 7.  ]
5.0
7.0
[3.14 7.  ]


Accessing values in a **matrix** numpy array is similar as well, but **instead of selecting elements we select rows**
- The index number now refers to the row number of the numpy array

In [ ]:
print(m)
print(m[0]) # first ROW
print(m[-1]) # last ROW
print(m[1:]) # slicing

[[11 12]
 [21 22]
 [31 32]]
[11 12]
[31 32]
[[21 22]
 [31 32]]


If instead you want to select elements in a two-dimensional matrix you need to specify a **second index number**:
```python
  m[i, j] # select the element at row index number i and column index number j
```

Remember, in Python we start counting from zero!

In [ ]:
print(m)
print(m[0, 1]) # SECOND element in the FIRST row
print(m[2, -1]) # LAST element in the THIRD row
print(m[2, :]) # ALL elements in the THIRD row
print(m[:, 0]) # ALL elements in the FIRST COLUMN

[[11 12]
 [21 22]
 [31 32]]
12
32
[31 32]
[11 21 31]


### Methods and mathematical operations

Numpy has implemented all sorts of methods that can be called on a numpy array:

> - **`mean`**: Returns the mean (average value)
> - **`var`**: Returns the variance (with `N` degrees of freedom by default)
> - **`max`**: Returns the maximum value
> - **`min`**: Returns the minimum value
> - **`std`**: Returns the standard deviation, that is, square root of the variance
> - **`sum`**: Returns the sum of the values

In [ ]:
print(v.var())
print(m.var())

2.4843555555555557
66.91666666666667


**Note**: By default the methods are applied to all elements, even for a two-dimensional **matrix**. If you want to get the mean per row, or per column, you need to specify the **`axis`** argument for the methods
>  - **`axis=0`**: Result **per column**, i.e. apply the method *over* the first dimension, that is the rows
>  - **`axis=1`**: Result **per row**, i.e. apply the method *over* the second dimension, that is the columns

In [ ]:
m.mean(axis=1)

array([11.5, 21.5, 31.5])

You can also apply mathematical operations **elementwise**
- This means that they are applied to each element of the numpy array.

**Note**: These manipulations **return a numpy array as the result but do not store the result**
- If you want to store the result you should assign it to a variable

In [ ]:
# Note: These operations do not change m
print(m + 2)
print(m * 2)
print(m / 2)
print(m**2)

[[13 14]
 [23 24]
 [33 34]]
[[22 24]
 [42 44]
 [62 64]]
[[ 5.5  6. ]
 [10.5 11. ]
 [15.5 16. ]]
[[ 121  144]
 [ 441  484]
 [ 961 1024]]


> **DO-IT**<br>
1. Add 10 to each element of the **`m`** variable and store it in a new variable called **`m_extra`**

In [ ]:
m_extra = m + 10
print(m_extra)

[[21 22]
 [31 32]
 [41 42]]


### Copying an array
Just like you can create new numpy arrays, you can also copy the contents of an existing one using the **`np.copy`** function:

In [ ]:
m_dup = np.copy(m)
print(m_dup)

[[11 12]
 [21 22]
 [31 32]]


### Making changes to an array
You can also make changes to a numpy array:

In [ ]:
print(m_dup)
m_dup[1] = m_dup[1] * 2 # multiplies the second row of m_dup x 2
print(m_dup)

[[11 12]
 [21 22]
 [31 32]]
[[11 12]
 [42 44]
 [31 32]]


**Note**: The original array **`m`**, of which **`m_dup`** is a copy is unchanged:

In [ ]:
print(m)

[[11 12]
 [21 22]
 [31 32]]


### Boolean comparison
In a similar way to the mathematical operators, we can apply the **comparison operators** to each element in a numpy array:

In [ ]:
m >= 22

array([[False, False],
       [False,  True],
       [ True,  True]])

You can use these **boolean** arrays to select values from a numpy array:
- **Note**: The values are returned as a one-dimensional **vector** numpy array

In [ ]:
m[m >= 22]

array([22, 31, 32])

Alternatively, you can use these **boolean** arrays to change values to be selected:

In [ ]:
m_dup = np.copy(m)
m_dup[m_dup >= 22] = 0 # all values that are >= 22 are set to 0
print(m_dup)

[[11 12]
 [21  0]
 [ 0  0]]


---
## **[C] Random sampling**

Besides the `numpy array`, NumPy contains some other functionality, such as **linear algebra functions**and functions for **random sampling** (that we will use later in the course).

The functionality to generate random numbers can be found in the **`np.random`** subpackage.


### Generator of random numbers
The first step is to create a new random number Generator using the **`default_rng`** function located in **`np.random`**:

In [ ]:
rng = np.random.default_rng()
# rng is our Generator now

For example, to generate a single Uniform random number between [0, 1) we can use the **`uniform`** method the Generator class:

In [ ]:
rng.uniform()

0.34035515523568904

> **DO-IT**<br>
> 1. Use the help function (`help(rng.uniform)`), or the ? syntax (`rng.uniform?`), or Google, to better understand the properties of the `rng.uniform` function.
> 2. Now, draw a Uniform random number that falls between 3.0 and 4.5

In [ ]:
#DO-IT

# 1.
#help(rng.uniform)
#rng.uniform?

# 2.
rng.uniform(low=3.0, high=4.5)

3.5716170157790303

If you want to generate more than one Uniform random number, you should specify the **`size`** argument of the **`rng.uniform`** function:

In [ ]:
# For example, to draw 10 random Uniform numbers:
rng.uniform(size=10)

array([0.31435398, 0.76503784, 0.29362225, 0.64693742, 0.99157927,
       0.22923054, 0.92472375, 0.16799036, 0.98407598, 0.01832078])

You can also generate a matrix (two-dimensional numpy array) by providing a **tuple** with sizes:

In [ ]:
rng.uniform(size=(4, 3))

array([[0.5455604 , 0.5850171 , 0.10916806],
       [0.93606559, 0.51895457, 0.962746  ],
       [0.80640273, 0.61207244, 0.08224731],
       [0.74198849, 0.54662974, 0.32439478]])

NumPy has functions to generate random numbers from all sorts of random distributions. For a complete list refer to the manual:<br> https://numpy.org/doc/stable/reference/random/generator.html#distributions

###Pseudo random

The random numbers that we generate are actually known as **pseudo random numbers** because they are not truly random. We can control the **state** of the random number generator such that our results become **reproducible**. **This is a very important property in data science, as you should be able to replicate your results**.

The way to control the **state** of the random number generator in NumPy is by specifing a **seed number** to the **`np.random.default_rng`** function:

In [ ]:
# WITHOUT specifying a seed:
rng_a = np.random.default_rng()
print(rng_a.uniform())

rng_b = np.random.default_rng()
print(rng_b.uniform())

# random numbers from rng_a and rng_b are not the same

0.8643511446758995
0.8127613911948399


In [ ]:
# WITH specifying a seed:

rng_a = np.random.default_rng(seed=744)
print(rng_a.uniform())

rng_b = np.random.default_rng(seed=744)
print(rng_b.uniform())

# random numbers from rng_a and rng_b now are the same

0.5557767730088412
0.5557767730088412


# **Python 3: pandas**


> - [A] Introduction
> - [B] `loc` and `iloc`


---
## **[A] Introduction**

We will import the **`pandas`** module using the alias **`pd`**, similar to what we have seen for `numpy`:

In [ ]:
import numpy as np
import pandas as pd

Today we will focus on the data structure in **`pandas`** called the **`DataFrame`**
> - The **`DataFrame`** in **`pandas`** is similar to the **data.frame in R**

### pandas DataFrame creation

We will first create a **`DataFrame`** using random numbers:
> - **Question**: What do we need to do to make sure that the random numbers are the same for all of us?

In [ ]:
my_rng_machine = np.random.default_rng()

random_data = my_rng_machine.standard_normal((5, 3))

print(random_data)

[[-0.2350546  -2.25270278 -2.19690417]
 [-1.31262927 -1.22033469  0.1682747 ]
 [-0.8952002  -0.4103195  -0.03888858]
 [ 1.2864492   0.83966544  0.11324581]
 [ 0.47557288  0.09421167  0.3000479 ]]


To create a **`DataFrame`** we need to call the **`pd.DataFrame`** function and provide data.

First, let us explore the help function for **`pd.DataFrame`**:

In [ ]:
pd.DataFrame?

We will now create our first Pandas DataFrame using our random data:

In [ ]:
df_00 = pd.DataFrame(random_data)
df_00 # note: print(df_00) does not have the nice interactive format!

,0,1,2
0,-0.235055,-2.252703,-2.196904
1,-1.312629,-1.220335,0.168275
2,-0.895200,-0.410319,-0.038889
3,1.286449,0.839665,0.113246
4,0.475573,0.094212,0.300048


You should notice a few things. First about the layout:
> - By just calling `df_00`, Jupyter Notebook actually created a nicely formatted DataFrame for us. It does not change the contents of the DataFrame, but makes it a little easier to visualize and interact with
> - Note that if we would have called `print(df_00)` this would not have happened.

Second, notice that pandas automatically created an **index** and **column labels** with default values:
> - **index**: Think of these as the **keys** to access the **rows** in a DataFrame
>  - In the example above the default index values are 0, 1, 2, 3, 4 because we have 5 rows
> - **column labels**: Think of these as the **keys** to access the **columns** in a DataFrame
>  - In the example above the default column labels are 0, 1, 2 because we have 3 columns

### Columns

We can define our own **column labels** by supplying a value to the **`columns`** argument in the call to **`pd.DataFrame`**, in addition to the data:

In [ ]:
# Note: Formatting on multiple lines is not necessary, but it makes the code
# easier to read
df_01 = pd.DataFrame(
    data=random_data,
    columns=['Col_1', 'Col 2', 'Column C'] # these are our column names
)
# The NUMBER of column names should match the NUMBER of columns in the data
df_01

,Col_1,Col 2,Column C
0,-0.235055,-2.252703,-2.196904
1,-1.312629,-1.220335,0.168275
2,-0.895200,-0.410319,-0.038889
3,1.286449,0.839665,0.113246
4,0.475573,0.094212,0.300048


If you start working with large datasets, you can always get a **list of column names** from the pandas DataFrame by accessing the **columns** attribute:

In [ ]:
print(df_01.columns)
print(type(df_01.columns))

Index(['Col_1', 'Col 2', 'Column C'], dtype='object')
<class 'pandas.core.indexes.base.Index'>


Notice that the `columns` attribute does not return a regular Python list, but a pandas `Index`
> A pandas `Index` is used to identify rows and columns in a pandas DataFrame

We can transform this `Index` to a "regular" Python `list` using the Python `list` function. The `list` function is a built-in function, simialr to the `len` function we worked with before:

In [ ]:
print(list(df_01.columns))
print(type(list(df_01.columns)))

['Col_1', 'Col 2', 'Column C']
<class 'list'>


You can select a single column from a DataFrame using the following syntax:
```python
  df['column_label']
```
where `df` is the DataFrame and column_label is the label of the column you want to select:

> **DO-IT**
> 1. Select Column C from `df_01`

In [ ]:
#DO-IT

# 1.
df_01['Column C']

0   -2.196904
1    0.168275
2   -0.038889
3    0.113246
4    0.300048
Name: Column C, dtype: float64

The return type of selecting a **single column** is a **pandas Series**:

In [ ]:
type(df_01['Column C'])

pandas.core.series.Series

> - **Note**: If you want to read more about the pandas Series consult its documentation: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html
> - Think of a Series as a **one-dimensional DataFrame**.
> - Make the analogy to a numpy array:
>  - A pandas Series corresponds to a vector (1-dimensional numpy array)
>  - A pandas DataFrame corresponds to a matrix (2-dimensional numpy array)

You can select multiple columns from a DataFrame by providing a **`list` of labels**:
```python
  df[['column_label', 'another_label']]
```

> **DO-IT**
> 1. Select both Column C and Col 2 (in that order) from `df_01`
> 2. Select both Col 2 and Col_1 (in that order) from `df_01`
> 1. Select Column C twice from `df_01`

In [ ]:
#DO-IT

# 1.
print('1.')
print(df_01[['Column C', 'Col 2']])

# 2.
print('2.')
print(df_01[['Col 2', 'Col_1']])

# 3.
print('3.')
print(df_01[['Column C', 'Column C']])

1.
   Column C     Col 2
0 -2.196904 -2.252703
1  0.168275 -1.220335
2 -0.038889 -0.410319
3  0.113246  0.839665
4  0.300048  0.094212
2.
      Col 2     Col_1
0 -2.252703 -0.235055
1 -1.220335 -1.312629
2 -0.410319 -0.895200
3  0.839665  1.286449
4  0.094212  0.475573
3.
   Column C  Column C
0 -2.196904 -2.196904
1  0.168275  0.168275
2 -0.038889 -0.038889
3  0.113246  0.113246
4  0.300048  0.300048


Note that the return type of selecting multiple columns is again a pandas DataFrame:

In [ ]:
type(df_01[['Column C', 'Col 2']])

pandas.core.frame.DataFrame

### Rows

We can define our own **index** for our pandas DataFrame by supplying a value to the **`index`** argument in the call to **`pd.DataFrame`**
> - This is in addition to the data and the column labels

In [ ]:
df_02 = pd.DataFrame(
    data=random_data,
    index=['a', 'e', 'i', 'j', 'o'], # these are our row names
    columns=['Col_1', 'Col 2', 'Column C']
)
# The NUMBER of row names should match the NUMBER of rows in the data
df_02

,Col_1,Col 2,Column C
a,-0.235055,-2.252703,-2.196904
e,-1.312629,-1.220335,0.168275
i,-0.895200,-0.410319,-0.038889
j,1.286449,0.839665,0.113246
o,0.475573,0.094212,0.300048


You can select multiple rows from a DataFrame using **slicing**:
```python
  df[start_index:stop_index] # the stop_index is again exclusive
```

> **DO-IT**
> 1. Select the second and third row from `df_02`
> 2. Select the last 2 rows from `df_02`
> 3. Select every other row from `df_02`, starting at the first row

In [ ]:
#DO-IT

# 1.
print('1.')
print(df_02[1:3])

# 2.
print('2.')
print(df_02[-2:]) # or print(df_02[3:5])

# 3.
print('3.')
print(df_02[0:5:2])

1.
      Col_1     Col 2  Column C
e -1.312629 -1.220335  0.168275
i -0.895200 -0.410319 -0.038889
2.
      Col_1     Col 2  Column C
j  1.286449  0.839665  0.113246
o  0.475573  0.094212  0.300048
3.
      Col_1     Col 2  Column C
a -0.235055 -2.252703 -2.196904
i -0.895200 -0.410319 -0.038889
o  0.475573  0.094212  0.300048


## [B] `loc` and `iloc`
The pandas DataFrame provides two additional ways to select values that are more flexible

> - **`loc`**: Select values based on their **labels**, i.e. using the **index**
> - **`iloc`**: Select values based on **position**, i.e. using the **row number** (and column) number

First, lets select rows with `loc`:

In [ ]:
df_02.loc[['a', 'e']] # this selects rows 'a' and 'e'

,Col_1,Col 2,Column C
a,-0.235055,-2.252703,-2.196904
e,-1.312629,-1.220335,0.168275


Next, lets select rows using `iloc`:

In [ ]:
df_02.iloc[0:2]

,Col_1,Col 2,Column C
a,-0.235055,-2.252703,-2.196904
e,-1.312629,-1.220335,0.168275


Both **`loc`** and **`iloc`** can also be used to select a subset of columns, as follows:

In [ ]:
# First, with loc
df_02.loc[:, ['Col_1', 'Column C']]

,Col_1,Column C
a,-0.235055,-2.196904
e,-1.312629,0.168275
i,-0.895200,-0.038889
j,1.286449,0.113246
o,0.475573,0.300048


In [ ]:
# Second, with iloc
df_02.iloc[:, [0, 2]]

,Col_1,Column C
a,-0.235055,-2.196904
e,-1.312629,0.168275
i,-0.895200,-0.038889
j,1.286449,0.113246
o,0.475573,0.300048


Naturally, a combination of selecting rows and columns is possible as well:

In [ ]:
df_02.loc[['e', 'o'], ['Col_1', 'Column C']]

,Col_1,Column C
e,-1.312629,0.168275
o,0.475573,0.300048


In [ ]:
df_02.iloc[[1, -1], [0, 2]]

,Col_1,Column C
e,-1.312629,0.168275
o,0.475573,0.300048


# Sample Questions

### Exercise - Writing functions in Python

Create a function called `greeting_func_00` that takes as input arguments two strings named `greeting` and `name`. The output of the function should print: "greeting, name."

> For example, if greeting="Hi" and name="Bob", the exact output should be "Hi, Bob."

**Hint**: It is likely that you will have a floating ".", for example that you will print "Hi, Bob ."
- You need to use the **`sep`** argument to the print function to change this behavior!
- You could use the Python documentation to figure this out: <br> https://docs.python.org/3/library/functions.html#print
- Sometimes that is too technical, and Googling will help, e.g.: <br> https://www.geeksforgeeks.org/python-sep-parameter-print/

In [ ]:
def greeting_func_00(greeting,name):
    print(greeting,',',name,'.',sep='')

greeting_func_00('Hi', 'Bob')





# def greeting_func_00(greeting, name):
#     print(greeting, ', ', name, '.', sep='')

# greeting_func_00('Hi', 'Bob')

Hi,Bob.


Now create a new function named `greeting_func_01` that has a third argument named `shout`. Again, if greeting="Hi" and name="Bob":
- If `shout=True` your function should now print: "HI, BOB!!!".
- If `shout=False`, your function should just print "Hi, Bob."

**Hint**: You most likely need to use one of the string methods to make strings uppercase. Refer to https://docs.python.org/3/library/stdtypes.html#string-methods (or search on Google)

In [ ]:
def greeting_func_01(greeting, name, shout):
    if shout:
        print(greeting.upper(), ', ', name.upper(), '!!!', sep='')
    else:
        print(greeting, ', ', name, '.', sep='')


greeting_func_01('Hi', 'Bob', shout=True)
greeting_func_01('Hi', 'Bob', shout=False)

HI, BOB!!!
Hi, Bob.


Finally, in our last function named `greeting_func_02` we want to use something that is called a **default argument**. This enables you to omit an argument to the function, in which case it will take on the **default value**.

> We actually have already seen this in action with the print function, where the `sep` argument by default is equal to a single space.

What I want you to do is make `shout=False` the default in `greeting_func_02`. You should do this in the definition of the function.

Now by calling `greeting_func_02('Hi', 'Bob')` the output printed should be `Hi, Bob.` If you call `greeting_func_02('Hi', 'Bob', shout=True)` the output printed should be `HI, BOB!!!`.

If you need more help with this, refer to: https://www.geeksforgeeks.org/default-arguments-in-python/

In [ ]:
def greeting_func_02(greeting, name, shout=False):
    if shout:
        print(greeting.upper(), ', ', name.upper(), '!!!', sep='')
    else:
        print(greeting, ', ', name, '.', sep='')



greeting_func_02('Hi', 'Bob', shout=True)
greeting_func_02('Hi', 'Bob')


SyntaxError: positional argument follows keyword argument (<ipython-input-30-270c49a0fae5>, line 11)

### Exercise - Conditional selection numpy array

In this exercise we will look at making conditional selections in numpy arrays.
> - **Note**: You should not be using for-loops in this exercise!

> **Instructions:**
> - Create a random number Generator and set the random seed to 744
> - Generate a `5 x 2` matrix of standard **Normal** random numbers using the `standard_normal` method of the Generator
> - Store the results in a variable **`m`**

The values you should generate should be equal to:
```
[[ 0.54043397  0.42074819]
 [ 0.70675007 -0.12944987]
 [ 1.46791925  1.17605899]
 [-0.10772229 -1.32806276]
 [ 0.39287602 -1.03068041]]
```



In [ ]:
import numpy as np
rng = np.random.default_rng(744)
m = rng.standard_normal(size=(5, 2))
print(m)

[[ 0.54043397  0.42074819]
 [ 0.70675007 -0.12944987]
 [ 1.46791925  1.17605899]
 [-0.10772229 -1.32806276]
 [ 0.39287602 -1.03068041]]


Create a numpy array named `m_bool` where its element is `True` if the corresponding element in `m < 0.0`, and `False` otherwise:

In [ ]:
m_bool = m < 0
m_bool




# m_bool = m < 0.0
# print(m_bool)

array([[False, False],
       [False,  True],
       [False, False],
       [ True,  True],
       [False,  True]])

How many values are smaller than 0.0?
> Hint: you should use the **`sum`** method or the **`np.sum`** function to count the number of `True` values in a numpy array

In [ ]:
print(m_bool.sum())



# print(m_bool.sum())

4


Select the values that are smaller than 0.0 from `m` using `m_bool`:

In [ ]:
m[m_bool]





# m[m_bool]

array([-0.12944987, -0.10772229, -1.32806276, -1.03068041])

Now take the sum of **each row** in `m`:

In [ ]:
print(m.sum())
print(m.sum(axis=1))





# m.sum(axis=1)

2.1088711550780346
[ 0.96118216  0.5773002   2.64397824 -1.43578506 -0.63780439]


Now take the mean of **each column** in `m`:

In [ ]:
m.mean(axis=0)





# m.mean(axis=0)

array([ 0.6000514 , -0.17827717])

### Exercise - Making changes to an array


1. Consider that you have stored a two-dimensional numpy array in a variable called m.

Consider the following:


*  Suppose m[1, 1] = -0.129
*  You run the following Python statement: m[:, 1] * 2
After that, you run the following Python statement: print(m[1, 1] > 1000)


What value will be printed?

In [ ]:
print(m)
print(m[1,1])
m[:, 1] * 2
print(m)
print(m[1, 1])
print(m[1, 1] > 1000)

[[ 0.54043397  0.42074819]
 [ 0.70675007 -0.12944987]
 [ 1.46791925  1.17605899]
 [-0.10772229 -1.32806276]
 [ 0.39287602 -1.03068041]]
-0.12944986816234996
[[ 0.54043397  0.42074819]
 [ 0.70675007 -0.12944987]
 [ 1.46791925  1.17605899]
 [-0.10772229 -1.32806276]
 [ 0.39287602 -1.03068041]]
-0.12944986816234996
False


### Exercise - DataFrame from a dictionary

In this exercise we will work with a DataFrame created from the dictionary below:

In [ ]:
some_data = populations = {
  'state': ['Florida','Texas','California','North Carolina', 'Virginia'],
  'population': [30.0, 28.3, 39.5, 10.3, 8.5]
}

Create a pandas DataFrame called `some_df` and set the index of the DataFrame equal to `['FL', 'TX', 'CA', 'NC', 'VA']`

In [ ]:
import pandas as pd
some_df = pd.DataFrame(
    data=some_data,
    index=['FL', 'TX', 'CA', 'NC', 'VA']
)
some_df

,state,population
FL,Florida,30.0
TX,Texas,28.3
CA,California,39.5
NC,North Carolina,10.3
VA,Virginia,8.5


Sort the DataFrame by population from large to small (in **descending** order) using the DataFrame method called **`sort_values`**. Refer to the help of this method to determine how to use its functionality.

**Hint**:
> - By default `sort_values` does **not change** your original dataframe
> - Instead, it only **returns** a copy of your sorted dataframe
> - You can change your original DataFrame by using the `inplace` argument from `sort_values`, or by saving the result in a variable
> - If you use the same variable name (`some_df`) you will overwrite your original DataFrame:

In [ ]:
some_df.sort_values?

# Using inplace=True
some_df.sort_values(by='population', ascending=False, inplace=True)

# # Or, by overwriting some_df
# some_df = some_df.sort_values(by='population', ascending=False)
some_df

,state,population
CA,California,39.5
FL,Florida,30.0
TX,Texas,28.3
NC,North Carolina,10.3
VA,Virginia,8.5


Again, using `loc` select the rows corresponding to Texas and North Carolina:

In [ ]:
some_df.loc[['TX','NC']]







# some_df.loc[['TX', 'NC']]

,state,population
TX,Texas,28.3
NC,North Carolina,10.3


Again, select the same states using `iloc`:

In [ ]:
some_df.iloc[[2,3]]







# some_df.iloc[[2, 3]]

,state,population
TX,Texas,28.3
NC,North Carolina,10.3


What method do you prefer to access rows from a DataFrame, `loc` or `iloc`? Why?

In [ ]:
# Personal preference (but I think this should be universal):
# loc
#
# Reason:
# We don't have to worry about the index numbers if the order of the
# rows change for some reason (like after we sorted the data)

Create a **boolean Series** and save the result in a variable named `exceeds_25` for which an element should be `True` if the population in a state is larger than 25 million, and `False` otherwise

**Hint**:
> - You can directly use comparison operators on a column of a DataFrame to create a boolean Series

In [ ]:
exceeds_25 = some_df['population'] > 25
exceeds_25









# exceeds_25 = some_df['population'] > 25
# exceeds_25

,population
CA,True
FL,True
TX,True
NC,False
VA,False


Using the `exceeds_25` variable and `loc` method, select the rows that correspond to states with more than 25 million residents:

In [ ]:
some_df[exceeds_25]










# some_df.loc[exceeds_25]

,state,population
CA,California,39.5
FL,Florida,30.0
TX,Texas,28.3
